lib/ace/transport_app.rb

```diff
@@ -33,6 +34,49 @@ def initialize(config = nil)
       super(nil)
     end
 
+    def self.init_puppet_target(device)
+      unless device['target']
+        ACE::Error.new("There was an error parsing the Puppet target. 'target' not found",
+                       'ace/puppet-target-error')
+        return nil
+      end
+      unless device['compiler']
+        ACE::Error.new("There was an error parsing the Puppet compiler details. 'compiler' not found",
+                       'ace/puppet-target-error')
+        return nil
+      end
+      unless device['compiler']['certname']
```

Comment:
You can save a lot of validation here by formulating the method as `init_puppet_target(certname, transport, target)`. Validation and extracting values from the API payload is more a responsibility of the endpoint rather than a business logic method.

Reply:
I guess I was a little confused as to who would call the function and what the input parameters should look like. I took the assumption that the output as per `execute_catalog` on https://github.com/willmeek/ace/blob/1de50b476296e0f5150ae748918a28f80a21733a/developer-docs/api.md would be the input